elastic/apm-server
# Data streams improvements

## Motivation/summary

Separate data streams support from existing index management and selector, and extend config validation to prevent mixing data streams and existing index management config (ILM, pipelines, template setup, index names, etc.)

When the server is running under Agent, it is now necessary for data streams to be enabled. We should inject the config (`-E apm-server.data_streams.enabled`) from the program spec.

Finally, we introduce system tests to show that with data streams enabled, the server works when given only limited privileges like those given by Fleet.

## Checklist

~- [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)~

I have considered changes for:
~- [ ] documentation~ (later)
~- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))~
~- [ ] telemetry~
~- [ ] Elasticsearch Service (https://cloud.elastic.co)~
~- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)~
~- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)~

## How to test these changes

- Run the server with `apm-server.data_streams.enabled=true`, should index into `traces-*` etc., with no pipeline specified at index time.
- Run `apm-server setup` with `apm-server.data_streams.enabled=true`, should fail.
- Run `apm-server setup --pipelines` with `apm-server.data_streams.enabled=true`, should fail.

## Related issues

https://github.com/elastic/apm-server/issues/4378
